compiler/rustc_passes/src/dead.rs

```diff
@@ -958,18 +998,36 @@ fn check_mod_deathness(tcx: TyCtxt<'_>, module: LocalModDefId) {
     let module_items = tcx.hir_module_items(module);
 
     for item in module_items.items() {
-        if let hir::ItemKind::Impl(impl_item) = tcx.hir().item(item).kind {
-            let mut dead_items = Vec::new();
-            for item in impl_item.items {
-                let def_id = item.id.owner_id.def_id;
-                if !visitor.is_live_code(def_id) {
+        let dead_codes = {
+            let filter_dead_item = |def_id| {
+                visitor.is_live_code(def_id).not().then(|| {
                     let name = tcx.item_name(def_id.to_def_id());
                     let level = visitor.def_lint_level(DEAD_CODE, def_id);
 
-                    dead_items.push(DeadItem { def_id, name, level })
-                }
+                    DeadItem { def_id, name, level }
+                })
+            };
+            match tcx.hir().item(item).kind {
+                ItemKind::Impl(impl_item) => impl_item
+                    .items
+                    .iter()
+                    // We already have diagnosed unused methods in traits
+                    .filter(|item| item.trait_item_def_id.is_none() || !matches!(item.kind, AssocItemKind::Fn { .. }))
+                    .filter_map(|item| filter_dead_item(item.id.owner_id.def_id))
+                    .collect(),
+                ItemKind::Trait(_, _, _, _, trait_items)
+                    // if we have diagnosed the trait, do not diagnose unused methods
+                    if live_symbols.contains(&item.owner_id.def_id) => trait_items
+                        .iter()
+                        // Only lints methods for now
+                        .filter(|trait_item| matches!(trait_item.kind, AssocItemKind::Fn { .. }))
+                        .filter_map(|trait_item| filter_dead_item(trait_item.id.owner_id.def_id))
+                        .collect(),
+                    _ => Vec::new(),
             }
```

Comment:
Nit: instead of looking at HIR, this can look at `def_kind(item)`, and check out `associated_item_def_ids` to get the nested items.